crates/oci-distribution/src/lib.rs

```diff
@@ -0,0 +1,174 @@
+    /// on other requests.
+    pub async fn auth(&mut self, host: String, secret: Option<String>) -> OciResult<()> {
+        let cli = reqwest::Client::new();
+        // The version request will tell us where to go.
+        let url = format!("https://{}/v2/", host);
+        let res = cli.get(&url).send().await?;
+        let disthdr = res.headers().get(reqwest::header::WWW_AUTHENTICATE);
+        if disthdr.is_none() {
+            // The Authenticate header can be set to empty string.
+            return Ok(());
+        }
+        // FIXME: There must be a more elegant way of passing a header from hyper into WwwAuthenticate.
+        let auth = WwwAuthenticate::parse_header(&disthdr.unwrap().to_str()?.to_string().into())?;
+        let challenge_opt = auth.get::<BearerChallenge>();
+        if challenge_opt.is_none() {
+            return Ok(());
+        }
+
+        let challenge = challenge_opt.as_ref().unwrap()[0].clone();
+        let realm = challenge.realm.unwrap();
+        let mut oauth = OauthClient::new(
+            "krustlet",
+            Url::parse(realm.as_str())?,
+            Url::parse(realm.as_str())?,
+        );
+        oauth.add_scope(challenge.scope.unwrap().as_str());
+        oauth.set_client_secret(secret.unwrap_or_else(|| "".to_owned()));
+
+        let token = oauth
+            .exchange_client_credentials()
+            .with_client(&cli)
+            .execute::<StandardToken>()
+            .await?;
+
+        self.token = Some(token);
+
+        Ok(())
+    }
+
+    pub async fn pull_manifest(&self, image_name: String) -> OciResult<String> {
+        // We unwrap right now because this try_from literally cannot fail.
+        let reference = Reference::try_from(image_name).unwrap();
+        let url = reference.to_v2_manifest_url();
+        let res = reqwest::get(&url).await?;
+        let status = res.status();
+        if res.status().is_client_error() {
+            // According to the OCI spec, we should see an error in the message body.
+            let err = res.json::<OciEnvelope>().await?;
+            // FIXME: This should not have to wrap the error.
+            return Err(anyhow::format_err!("{}", err.errors[0]));
+        } else if status.is_server_error() {
+            return Err(anyhow::format_err!("Server error at {}", url));
+        }
+        let text = res.text().await?;
+        Ok(text)
+    }
+}
+
+#[derive(Clone)]
+struct BearerChallenge {
+    pub realm: Option<String>,
+    pub service: Option<String>,
+    pub scope: Option<String>,
+}
+
+impl Challenge for BearerChallenge {
+    fn challenge_name() -> &'static str {
+        "Bearer"
+    }
+    fn from_raw(raw: RawChallenge) -> Option<Self> {
+        match raw {
+            RawChallenge::Token68(_) => None,
+            RawChallenge::Fields(mut map) => Some(BearerChallenge {
+                // NB (mpb): This follows the `remove` pattern from the existing
```

Comment:
This is to make ownership of the items that were in the map. Otherwise you'd need to clone.

Reply:
Oh! That totally makes sense!